### ISSUE TITLE
Joint distribution and polynomial expansion

### ISSUE THREAD

**Issue by jigarparekh279** [jonathf/chaospy#228]
Hi Jonathan,

I am trying to create a joint distribution from two different sets of Iids and then generate the polynomial expansion as follows-

``` python
import chaospy as cp

dist_ij   = cp.J(cp.Normal(0,1), cp.Normal(0,1), cp.Normal(0,1))
dist_ii   = cp.J(cp.Gamma(1), cp.Gamma(1))

dist      = cp.J(*dist_ij, *dist_ii)
polyExpsn = cp.orth_ttr(3, dist, retall=False, normed=True)  
```

The above works fine. But when I try to change it to -

```python
dn = 3 #10
dg = 2 #4

dist_ij   = cp.Iid(cp.Normal(0,1), dn)
dist_ii   = cp.Iid(cp.Gamma(1), dg)

dist      = cp.J(*dist_ij, *dist_ii)
polyExpsn = cp.orth_ttr(3, dist, retall=False, normed=True)  
```
it returns an error -

```python

  File "/home/jigar/anaconda3/envs/jigarEnv/lib/python3.7/site-packages/chaospy/distributions/operators/joint.py", line 221, in _ttr
    kloc_ = kloc[idx].reshape(1)

IndexError: index 4 is out of bounds for axis 0 with size 2
```
Could you please help me out.

**Comment by jonathf** [comment]
I can not reproduce the error on my machine, but the Iid wrapper was updated not long ago, so I imagine you are not using up-do-date version of chaospy.

Can you confirm that you are using version 3.2.9?
If not, can you update to 3.2.9 and see if the problem still persist?

**Comment by jigarparekh279** [comment]
Thanks for the reply. I updated chaospy to version 3.2.9 and it works. However, if I increase the number of dimensions -
```py
import chaospy as cp

dn = 15
dg = 3

dist_ij   = cp.Iid(cp.Normal(0,1), dn)
dist_ii   = cp.Iid(cp.Gamma(1), dg)

dist      = cp.J(*dist_ij, *dist_ii)
polyExpsn = cp.orth_ttr(3, dist, retall=False, normed=True)
```
I run into memory issues -
```py
Unable to allocate array with shape (16777216, 12) and data type int64
```
This wasn't a problem when I was using an older version (3.0.11).

**Comment by jonathf** [comment]
I like that you are stretching the limits for what chaospy is designed to do. Thunmbs up for that.

I have made a fix to the numpoly backend. Update the dependency to version 0.1.14 `pip install --upgrade numpoly`, and you should be able to do `dn = 25` again I think.

**Comment by jigarparekh279** [comment]
Hi, thanks for the thumbs up and for upgrading ```numpoly```. The previous issue is resolved but I ran into a new one while computing the inner products of three or more polynomials using outer product -

```py
import chaospy as cp

dn = 15
dg = 3

dist_ij   = cp.Iid(cp.Normal(0,1), dn)
dist_ii   = cp.Iid(cp.Gamma(1), dg)

dist      = cp.J(*dist_ij, *dist_ii)
polyExpsn = cp.orth_ttr(3, dist, retall=False, normed=True, cross_truncation=0.1)

phi2 = cp.outer(polyExpsn, polyExpsn)
phi3 = cp.outer(polyExpsn, polyExpsn, polyExpsn)
phi4 = cp.outer(polyExpsn, polyExpsn, polyExpsn, polyExpsn)

Cij   = cp.E(phi2, dist)
Cijk  = cp.E(phi3, dist)
Cijkl = cp.E(phi4, dist)
```
this is the error I get -
```py
ValueError: non-broadcastable output operand with shape (25,) doesn't match the broadcast shape (25,25)
```
Again this wasn't an issue in the older version (3.0.11). Also, I see that routine for ```cp.outer()``` has changed. Can these inner products can be computed any other way (except looping over each polynomial index in the product)?

**Comment by jonathf** [comment]
Outer has indeed changed, and it was deliberate.

The new backend for the polynomials [numpoly](https://github.com/jonathf/numpoly) uses subclassing of ``numpy.ndarray`` as foundation to create a high level of compatibility. For example, the following two commands are equivalent:
```python
numpy.outer(polyExpsn, polyExpsn)
chaospy.outer(polyExpsn, polyExpsn)
```

To achieve this (for this function and many others), some of the call signatures had to be changed to match their numpy counterparts.

In other words, the code you are running will not work, and to keep compatibility with numpy will make it unfeasible to have custom signatures deviating from the what numpy is providing.

However, in your case, the following should work:
```python
phi2 = cp.outer(polyExpsn, polyExpsn)
phi3 = cp.outer(polyExpsn, cp.outer(polyExpsn, polyExpsn)).reshape(N, N, N)
phi4 = cp.outer(polyExpsn, cp.outer(polyExpsn, cp.outer(polyExpsn, polyExpsn))).reshape(N, N, N, N)
```
Hope this helps.

**Comment by jigarparekh279** [comment]
Thanks for the reply. I understand the need to change ```choaspy.outer()``` routine. Anyway, the alternative you suggest also works until I run out of memory on my machine. But then I can always use loops to compute the inner products when the number of dimensions get really large. I'll close the issue now.
